libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp

```diff
@@ -0,0 +1,312 @@
+	case Instruction::MLOAD:
+	case Instruction::MSTORE:
+	case Instruction::MSTORE8:
+	case Instruction::SLOAD:
+	case Instruction::SSTORE:
+	case Instruction::PC:
+	case Instruction::MSIZE:
+	case Instruction::GAS:
+	case Instruction::LOG0:
+	case Instruction::LOG1:
+	case Instruction::LOG2:
+	case Instruction::LOG3:
+	case Instruction::LOG4:
+	case Instruction::TLOAD:
+	case Instruction::TSTORE:
+		return ImpureBuiltinEncountered();
+	// --------------- calls ---------------
+	case Instruction::CREATE:
+	case Instruction::CREATE2:
+	case Instruction::CALL:
+	case Instruction::CALLCODE:
+	case Instruction::DELEGATECALL:
+	case Instruction::STATICCALL:
+	case Instruction::RETURN:
+	case Instruction::REVERT:
+	case Instruction::INVALID:
+	case Instruction::SELFDESTRUCT:
+		return ImpureBuiltinEncountered();
+
+	case Instruction::POP:
+		return EvaluationOk();
+	// --------------- invalid in strict assembly ---------------
+	case Instruction::JUMP:
+	case Instruction::JUMPI:
+	case Instruction::JUMPDEST:
+	case Instruction::PUSH0:
+	case Instruction::PUSH1:
+	case Instruction::PUSH2:
+	case Instruction::PUSH3:
+	case Instruction::PUSH4:
+	case Instruction::PUSH5:
+	case Instruction::PUSH6:
+	case Instruction::PUSH7:
+	case Instruction::PUSH8:
+	case Instruction::PUSH9:
+	case Instruction::PUSH10:
+	case Instruction::PUSH11:
+	case Instruction::PUSH12:
+	case Instruction::PUSH13:
+	case Instruction::PUSH14:
+	case Instruction::PUSH15:
+	case Instruction::PUSH16:
+	case Instruction::PUSH17:
+	case Instruction::PUSH18:
+	case Instruction::PUSH19:
+	case Instruction::PUSH20:
+	case Instruction::PUSH21:
+	case Instruction::PUSH22:
+	case Instruction::PUSH23:
+	case Instruction::PUSH24:
+	case Instruction::PUSH25:
+	case Instruction::PUSH26:
+	case Instruction::PUSH27:
+	case Instruction::PUSH28:
+	case Instruction::PUSH29:
+	case Instruction::PUSH30:
+	case Instruction::PUSH31:
+	case Instruction::PUSH32:
+	case Instruction::DUP1:
+	case Instruction::DUP2:
+	case Instruction::DUP3:
+	case Instruction::DUP4:
+	case Instruction::DUP5:
+	case Instruction::DUP6:
+	case Instruction::DUP7:
+	case Instruction::DUP8:
+	case Instruction::DUP9:
+	case Instruction::DUP10:
+	case Instruction::DUP11:
+	case Instruction::DUP12:
+	case Instruction::DUP13:
+	case Instruction::DUP14:
+	case Instruction::DUP15:
+	case Instruction::DUP16:
+	case Instruction::SWAP1:
+	case Instruction::SWAP2:
+	case Instruction::SWAP3:
+	case Instruction::SWAP4:
+	case Instruction::SWAP5:
+	case Instruction::SWAP6:
+	case Instruction::SWAP7:
+	case Instruction::SWAP8:
+	case Instruction::SWAP9:
+	case Instruction::SWAP10:
+	case Instruction::SWAP11:
+	case Instruction::SWAP12:
+	case Instruction::SWAP13:
+	case Instruction::SWAP14:
+	case Instruction::SWAP15:
+	case Instruction::SWAP16:
+	{
+		yulAssert(false, "");
+		return EvaluationOk(0);
+	}
+	}
+
+	yulAssert(false, "Unknown instruction with opcode " + std::to_string(static_cast<uint8_t>(_instruction)));
+	return EvaluationOk(0);
```

Comment:
```suggestion
	case Instruction::SWAP16:
	{
		yulAssert(false);
		util::unreachable();
	}
	}

	yulAssert(false);
	util::unreachable();
```

Reply:
Thank you for pointing out the `util::unreachable`. I actually don't know about it.

However, I think the assert with error message should be used. Currently all the instruction are checked with a `switch` statement, which is not guaranteed to be exhaustive check at compile time. I am not sure if there is a way to ensure its correctness at compile time. For the runtime check, I think it should fail loudly. When failure happens, there must be a new instruction. It should then be added to this function.

Reply:
> which is not guaranteed to be exhaustive check at compile time

If you do not include a `default` case, most compilers (at least GCC and Clang) will warn when you don't handle all cases. We use warnings-as-errors by default so it will just not compile instead of failing at runtime. So correctness in fact *is* ensured here at compilation time already.

Actually, the compiler should even figure out on its own that any code after the `switch` is unreachable, and usually it does. `unreachable()` is just our workaround for a GCC bug (see #13087) that makes it fail to do so in Debug mode.

Reply:
I changed to `util::unreachable()`. Thank you for the insight about the workaround! :pray: